lib/translator/write.cc

```diff
@@ -0,0 +1,217 @@
+              .nsid = nsid};
+
+  nvme_cmd.dptr.prp.prp1 = allocation.data_addr;
+
+  return status_code;
+}
+
+StatusCode Write(bool fua, uint8_t wrprotect, uint32_t transfer_length,
+                 nvme::GenericQueueEntryCmd& nvme_cmd, Allocation& allocation,
+                 uint32_t nsid, uint32_t page_size, uint32_t lba_size) {
+  if (transfer_length == 0) {
+    DebugLog("NVMe read command does not support transfering zero blocks\n");
+    return StatusCode::kNoTranslation;
+  }
+
+  transfer_length &= 0xffff;  // truncate to 16 bits
+  StatusCode status_code =
+      LegacyWrite(nvme_cmd, allocation, nsid,
+                  GetTransferLengthPages(transfer_length, page_size, lba_size));
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  uint8_t pr_info;
+  status_code = BuildPRInfo(wrprotect, pr_info);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+  nvme_cmd.cdw[2] = (transfer_length - 1 | pr_info << 26 | fua << 30);
+
+  return status_code;
+}
+
+}  // namespace
+
+StatusCode Write6ToNvme(Span<const uint8_t> scsi_cmd,
+                        nvme::GenericQueueEntryCmd& nvme_cmd,
+                        Allocation& allocation, uint32_t nsid,
+                        uint32_t page_size, uint32_t lba_size) {
+  scsi::Write6Command write_cmd = {};
+  if (!ReadValue(scsi_cmd, write_cmd)) {
+    DebugLog("Malformed Write6 Command");
+    return StatusCode::kInvalidInput;
+  }
+
+  // A TRANSFER LENGTH field set to zero specifies that 256 logical blocks shall
+  // be written. Any other value specifies the number of logical blocks that
+  // shall be written (Section 3.59) of
+  // https://www.seagate.com/files/staticfiles/support/docs/manual/Interface%20manuals/100293068j.pdf
+  uint16_t updated_transfer_length =
+      (write_cmd.transfer_length == 0) ? 256 : write_cmd.transfer_length;
+  StatusCode status_code = LegacyWrite(
+      nvme_cmd, allocation, nsid,
+      GetTransferLengthPages(updated_transfer_length, page_size, lba_size));
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  // cdw10 Starting lba bits 31:00
+  uint32_t host_endian_lba = (write_cmd.logical_block_address_1 << 16) +
+                             ntohs(write_cmd.logical_block_address);
+
+  nvme_cmd.cdw[0] = htoll(host_endian_lba);
+  nvme_cmd.cdw[2] = htoll(static_cast<uint32_t>(updated_transfer_length - 1));
+  return status_code;
+}
+
+StatusCode Write10ToNvme(Span<const uint8_t> scsi_cmd,
+                         nvme::GenericQueueEntryCmd& nvme_cmd,
+                         Allocation& allocation, uint32_t nsid,
+                         uint32_t page_size, uint32_t lba_size) {
+  scsi::Write10Command write_cmd = {};
+  if (!ReadValue(scsi_cmd, write_cmd)) {
+    DebugLog("Malformed Write10 Command");
+    return StatusCode::kInvalidInput;
+  }
+
+  StatusCode status_code =
+      Write(write_cmd.fua, write_cmd.wr_protect, write_cmd.transfer_length,
```

Comment:
Here transfer length is 16 bits and should be converted to host endian before using it in `GetTransferLengthPages`
```suggestion
      Write(write_cmd.fua, write_cmd.wr_protect, ntohs(write_cmd.transfer_length),
```

Reply:
https://github.com/googleinterns/scsi2nvme/pull/128#discussion_r465369072